### ISSUE TITLE
Training problem

### ISSUE THREAD

**Issue by wangyu1213888** [lxa9867/R2VOS#5]
Could you please tell me what parameters need to be specified for training? I'm referring to running the command python3 main.py with options like --with_box_refine, --binary, --freeze_text_encoder, --ngpu 1, --use_cycle, --mix_query, --use_fg_contra，--neg_cls, --use_cls, --output_dir=ytvos_dirs/resnet50, and --pretrained_weights=pretrained_weights/r50_pretrain.pth. What would the complete command look like? 
I always make the following mistakes when training：
Loss is nan, stopping training
{'loss_ce': tensor(0.2781, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor(0.8912, device='cuda:0', grad_fn=<DivBackward0>), 'loss_giou': tensor(1.0353, device='cuda:0', grad_fn=<DivBackward0>), 'loss_mask': tensor(0.0142, device='cuda:0', grad_fn=<DivBackward0>), 'loss_dice': tensor(0.1781, device='cuda:0', grad_fn=<DivBackward0>), 'loss_cycle_dist': tensor(nan, device='cuda:0', grad_fn=<SmoothL1LossBackward0>), 'loss_cycle_angle': tensor(0., device='cuda:0', grad_fn=<SmoothL1LossBackward0>), 'loss_cycle_mse': tensor(16.8410, device='cuda:0', grad_fn=<MeanBackward0>), 'loss_cycle_contrastive': tensor(0., device='cuda:0', grad_fn=<ClampBackward1>), 'loss_cycle_cls': tensor(0.6948, device='cuda:0',
       grad_fn=<BinaryCrossEntropyWithLogitsBackward0>), 'loss_fg_contra': tensor(1.2444, device='cuda:0', grad_fn=<NllLossBackward0>), 'loss_VQ': tensor(2.7403, device='cuda:0', grad_fn=<AddBackward0>), 'loss_ce_0': tensor(0.2831, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox_0': tensor(0.8821, device='cuda:0', grad_fn=<DivBackward0>), 'loss_giou_0': tensor(1.0914, device='cuda:0', grad_fn=<DivBackward0>), 'loss_mask_0': tensor(0.0197, device='cuda:0', grad_fn=<DivBackward0>), 'loss_dice_0': tensor(0.1742, device='cuda:0', grad_fn=<DivBackward0>), 'loss_ce_1': tensor(0.2792, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox_1': tensor(0.8986, device='cuda:0', grad_fn=<DivBackward0>), 'loss_giou_1': tensor(1.0750, device='cuda:0', grad_fn=<DivBackward0>), 'loss_mask_1': tensor(0.0141, device='cuda:0', grad_fn=<DivBackward0>), 'loss_dice_1': tensor(0.1859, device='cuda:0', grad_fn=<DivBackward0>), 'loss_ce_2': tensor(0.2896, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox_2': tensor(0.8999, device='cuda:0', grad_fn=<DivBackward0>), 'loss_giou_2': tensor(1.0165, device='cuda:0', grad_fn=<DivBackward0>), 'loss_mask_2': tensor(0.0141, device='cuda:0', grad_fn=<DivBackward0>), 'loss_dice_2': tensor(0.1790, device='cuda:0', grad_fn=<DivBackward0>)}

I wonder if this is because I am using non-distributed related.

**Comment by lxa9867** [comment]
It is recommended to borrow the training script from referformer. For the cycle loss, I think it must be in the ddp training.

```
    def loss_cycle(self, outputs, targets, indices, num_boxes):
        src_sentence_embedding = outputs['pseudo_sentence_feature'].squeeze(0)  # BxC
        tgt_sentence_embedding = outputs['sentence_feature']

        # gather all
        src_sentence_embedding_list = [torch.zeros_like(src_sentence_embedding) for _ in range(dist.get_world_size())]
        tgt_sentence_embedding_list = [torch.zeros_like(tgt_sentence_embedding) for _ in range(dist.get_world_size())]
        dist.all_gather(src_sentence_embedding_list, src_sentence_embedding)
        dist.all_gather(tgt_sentence_embedding_list, tgt_sentence_embedding)
```

**Comment by wangyu1213888** [comment]
referformer's training script specifies the parameters as 
```
python3 -m torch.distributed.launch --nproc_per_node=${GPUS_PER_NODE} --master_port=${PORT} --use_env \
main.py --with_box_refine --binary --freeze_text_encoder \
--epochs 6 --lr_drop 3 5 \
--output_dir=${OUTPUT_DIR} --pretrained_weights=${PRETRAINED_WEIGHTS} ${PY_ARGS} 
```
We don't use the loss function you mentioned in your paper, I think it should be 
```
python3 -m torch.distributed.launch --nproc_per_node=${GPUS_PER_NODE} --master_port=${PORT} --use_env \
main.py --with_box_refine --binary --freeze_text_encoder  --use_cycle --add_negative --mix_query  --neg_cls  --freeze_quantitizer --use_fg_contra --quantitize_query --bert_cycle --use_cls \
--epochs 12 --lr_drop 8 10 \
--output_dir=${OUTPUT_DIR} --pretrained_weights=${PRETRAINED_WEIGHTS} ${PY_ARGS} 
```
I was wondering if the script I set up is missing any parameters for your code

**Comment by lxa9867** [comment]
Yes, I think you are correct. Several minor changes are required to use the main.py in referformer.
